### PR TITLE
[server] incremental workspaces (1 of 2)

### PR DESCRIPTION
**Pull request by svenefftinge**
# Attention
> **Please review and approve #14461 which includes these changes**

## Description
This change adds a project preference that allows to generally use previous prebuilds and incrementally update the workspace when creating a workspace:

Screenshot 2022-11-06 at 21 38 35

## Related Issue(s)

See #12582

## How to test

- create a team and a project
- enable the project setting
- start a prebuild on that project
- start a workspace while a prebuild is running and see that you get a workspace starting from git.
- wait for the prebuild to finish
- add a commit on top so that a prebuild is running again
- start a workspace while a prebuild is running and see that you get a workspace starting from git.
- change the new preference in project settings
- start a workspace while the prebuild is running and see that you get a workspace based on the previous prebuild.

## Release Notes

```release-note
Introduces new project setting to use old prebuilds on workspace start
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: components/dashboard/src/projects/Project.tsx
Comment (svenefftinge) [inline components/dashboard/src/projects/Project.tsx:381]: We should always start on the branch context here, no matter what prebuild state exists

File: components/dashboard/src/settings/selectClass.tsx
Comment (svenefftinge) [inline components/dashboard/src/settings/selectClass.tsx:60]: Moved to the call site

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (svenefftinge) [inline components/dashboard/src/start/CreateWorkspace.tsx:559]: polling here every 2 secs is much simpler (and actually works). Listening to events has a problem where we could have missed the 'stopped' event and then hang forever

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (svenefftinge) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1069]: everything deleted below is no longer needed because we are now polling for prebuild finish. See above ☝️

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (easyCZ) [reply components/dashboard/src/start/CreateWorkspace.tsx:559]: Love this! I'm a huge fan of this simpler approach, and it also aligns much better with where we are looking to go with Pub API.

**Review by easyCZ — APPROVED** [review]
LGTM. I would recommend splitting up the polling changes from the rest of the incremental workspaces feature.

/hold if you'd like a review from someone else also, I may not have 100% of the context here

**Comment by geropl** [comment]
Starting to review now.

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/CreateWorkspace.tsx:268]: :+1:

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/CreateWorkspace.tsx:154]: :+1:

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [reply components/dashboard/src/start/CreateWorkspace.tsx:559]: I'm fine with either way. Can we use [repeat](https://github.com/gitpod-io/gitpod/blob/2f0b9462b0451f0974434fa36bd757420ef04e34/components/gitpod-protocol/src/util/repeat.ts#L17) or any other mechnaims that ensures we cannot pile up requests here? :cloud:

File: components/gitpod-protocol/src/gitpod-service.ts
Comment (geropl) [inline components/gitpod-protocol/src/gitpod-service.ts:423]: :orange_heart:

File: components/gitpod-protocol/src/protocol.ts
Comment (geropl) [inline components/gitpod-protocol/src/protocol.ts:215]: Would it make sense to encapsulate these options in a sub-shape ala `workspaceStartOptions` or so? That would make some of the other code shorter as well, and help keeping this one clean(er).

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1071]: :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1154]: :+1:

File: components/dashboard/src/projects/Project.tsx
Comment (geropl) [reply components/dashboard/src/projects/Project.tsx:381]: Why? Because we assume everyone will enable incremental workspaces (which results in the best Git state)? :thinking:

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1001]: This slight change in behavior (never starting based on previous prebuilds for OpenPrebuildContext) makes sense :100:

**Comment by geropl** [comment]
Code LGTM, starting testing now.

**Comment by geropl** [comment]
Because of the missing preview env and failing build I took the liberty to rebase on latest main with the goal to test...

File: components/dashboard/src/projects/Project.tsx
Comment (geropl) [reply components/dashboard/src/projects/Project.tsx:381]: Ah sorry: Noticed this is the "Branches" view. I agree. :+1:

**Comment by geropl** [comment]
@svenefftinge Can you explain why you think exposing the `ignoreRunningWorkspaceOnSameCommit` option make sense? To me it feels like an implementation detail of the API, which is offers a nice catch-all for potentially bad UX flows. Or am I missing sth? :thinking: 
image

**Comment by geropl** [comment]
I tested back and forth, and it seems to mostly just work (TM) :slightly_smiling_face: 
Except when "Never wait for running prebuilds" = false && "Allow incremental workspaces" = true I never get to see the prebuild waiting screen.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1002]: Would it make sense to change this to `prebuiltWorkspace?.state !== "available" && ignoreRunningPrebuild && allowUsingPreviousPrebuilds`?
Otherwise exposing the `ignoreRunningPrebuild` option to users makes less sense, because it is not independent of `allowUsingPreviousPrebuild`, and we would have to explain that connection. :confused:

**Comment by geropl** [comment]
@svenefftinge Given the last two comments [[1](https://github.com/gitpod-io/gitpod/pull/14303#issuecomment-1303658497), [2](https://github.com/gitpod-io/gitpod/pull/14303#issuecomment-1303530355)] I think it would make sense to just expose a single option: the incremental workspace one. Motivation: the other options are hard to explain, add less additional value, and are not independent.
WDYT?

BTW I'd love to approve a PR that "just" replaces the mode with the option fields, that alone is :orange_heart: :upside_down_face:

**Comment by svenefftinge** [comment]
Thanks Gero! I've worked a bit on this with @mbrevoort as well and we agreed that we should only have the `incremental prebuild` setting. I'm moving this to the project level also, to make it easier enabling it for everyone. 
We also believe that waiting for non-available prebuilds is never useful, so would like to remove this path entirely. WDYT?

**Comment by svenefftinge** [comment]
> @svenefftinge Can you explain why you think exposing the `ignoreRunningWorkspaceOnSameCommit` option make sense?
Yeah, it doesn't make sense. I think I did it because I could 😅 Have removed it.

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/CreateWorkspace.tsx:270]: Can we put this behind a feature flag, just to decouple it from the rollout of the "incremental workspace" feature?

**Comment by geropl** [comment]
> we agreed that we should only have the incremental prebuild setting. I'm moving this to the project level also
 
:100: d'accord: project level also is nice because it_s a property of the project whether it supports incremental build

> We also believe that waiting for non-available prebuilds is never useful

@svenefftinge I believe the same, but would be a bit more cautios with rolling out this change. Can we put this behind a feature flag? Just in case there are issues, e.g. unexpecte UX flow bugs etc.? Also, it would feel way better to keep the rollout separate from the other change. :slightly_smiling_face:

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/CreateWorkspace.tsx:270]: I could make the removal of `runningWorkspacePrebuild` a separate PR, that we merge first. But I'm not sure that buys us a lot. Would be super helpful to dive into the uncertainty a bit more and figure out if there is something we missed. I can't think of any issues that the removal of this feature could have on existing workflows or deployment.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:104]: **suggestion**: Things to consider before merging:
1. From what I understand, the persistence feature below affects both prebuilds and workspaces, but is it valuable or helpful to surface this to users?
2. Could it make sense to bundle this setting with the PVC setting below for simplicity, especially since the setting is currently not working as expected (https://github.com/gitpod-io/gitpod/issues/14159)?
3. Using an alternative description to make it easier for users to understand the benefit of this option.

| BEFORE | AFTER |
|-|-|
| project-settings-before | project-settings-after |

**Review by gtsiolis — COMMENTED** [review]
Hey everyone! Saw this new setting a left a drive-by comment below which can be also tackled in a follow-up issue. 😁

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [reply components/dashboard/src/start/CreateWorkspace.tsx:270]: At the core, the concern I have is: What happens to a high-frequency repo, that did not enable incremental workspaces yet, once we roll this out? To me it feels they'd just never benefit from Prebuilds anymore. Until they understand that there is Incremental Workspaces, and it works for them. This feels like a bad spot for us to be in, and it feels unnecessary as well.

Alternatively, we could leave the screen as-is, maybe tied to `!opts.incrementalWorkspace`. That way, people can still try to manually fall back to "Use last successful prebuild" (current code) if they want to.
Plus it gives us an obvious spot where to out the advertisement for the "Incremental Workspaces" option, which people can learn about and "migrate" to. :lotus_position:

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (svenefftinge) [reply components/dashboard/src/projects/ProjectSettings.tsx:104]: Let's fine tune the preferences on this follow up pr https://github.com/gitpod-io/gitpod/pull/14461

Comment (gtsiolis) [reply components/dashboard/src/projects/ProjectSettings.tsx:104]: OK!

**Comment by svenefftinge** [comment]
let's go for https://github.com/gitpod-io/gitpod/pull/14461

**Pull request closed without merging**
